examples/native_training/main.c

```diff
@@ -0,0 +1,260 @@
+    fprintf(stdout, "Configuring IREE runtime instance and '%s' device\n",
+            driver_name.data);
+    status = iree_runtime_instance_create(&instance_options,
+                                          iree_allocator_system(), &instance);
+  }
+  // TODO(#5724): move device selection into the compiled modules.
+  iree_hal_device_t* device = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_runtime_instance_try_create_default_device(
+        instance, driver_name, &device);
+  }
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Session configuration (one per loaded module to hold module state).
+  iree_runtime_session_options_t session_options;
+  iree_runtime_session_options_initialize(&session_options);
+  iree_runtime_session_t* session = NULL;
+  if (iree_status_is_ok(status)) {
+    fprintf(stdout, "Creating IREE runtime session\n");
+    status = iree_runtime_session_create_with_device(
+        instance, &session_options, device,
+        iree_runtime_instance_host_allocator(instance), &session);
+  }
+  iree_hal_device_release(device);
+
+  if (iree_status_is_ok(status)) {
+    fprintf(stdout, "Loading bytecode module at '%s'\n",
+            bytecode_module_path.data);
+    status = iree_runtime_session_append_bytecode_module_from_file(
+        session, bytecode_module_path.data);
+  }
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  if (iree_status_is_ok(status)) {
+    fprintf(stdout, "Training...\n");
+    print_state(state);
+    for (int i = 0; i < 10; i++) {
+      status = train(session, state);
+      print_state(state);
+      if (!iree_status_is_ok(status)) {
+        break;
+      }
+    }
+  }
+  //===-------------------------------------------------------------------===//
+
+  //===-------------------------------------------------------------------===//
+  // Cleanup.
+  iree_runtime_session_release(session);
+  iree_runtime_instance_release(instance);
+  //===-------------------------------------------------------------------===//
+
+  return status;
+}
+
+int main(int argc, char** argv) {
+  // Parse args
+  if (argc < 2) {
+    fprintf(stderr,
+            "Usage: native-training </path/to/native_training.vmfb> "
+            "[<driver_name>]\n");
+    fprintf(stderr, "  (See the README for this sample for details)\n ");
+    return -1;
+  }
+  iree_string_view_t bytecode_module_path = iree_make_cstring_view(argv[1]);
+  iree_string_view_t driver_name;
+  if (argc >= 3) {
+    driver_name = iree_make_cstring_view(argv[2]);
+  } else {
+    driver_name = iree_make_cstring_view("local-sync");
+  }
+
+  // Run training
+  struct State state = {
+      {4.0f, 4.0f, 5.0f},  // w
+      {2.0f},              // b
+      {1.0f, 1.0f, 1.0f},  // X
```

Comment:
Doesn't have to be fixed in this patch, but to make this example actually realistic, I think it is important that it actually trains on a decent amount of data. To do that, I would recommend that you implement a C function that generates a "large" volume of data for X and y with values of W and b that the model doesn't see. And I would invoke the training procedure on that large volume of data and actually assert that the values of W and b that the model learns end up "close" to the values that the model doesn't see.